Crash: release called twice for the same resource

### Haraka version
 2.8.8

### Observed behavior
```
[ERROR] [806DE749-F0A2-4BA7-9F4D-74FCEDCBAE99.1.1] [outbound] Ongoing connection failed to cli.ent.ip.addr:25 : Error: write after end
[INFO] [-] [core] [outbound] [outbound::25:cli.ent.ip.addr:undefined:50] dispense() clients=0 available=1
[INFO] [806DE749-F0A2-4BA7-9F4D-74FCEDCBAE99.1.1] [outbound] Temp failing 1472505160080_1_5187_8878.relay for 128 seconds: Tried all MXs
[NOTICE] [806DE749-F0A2-4BA7-9F4D-74FCEDCBAE99.1.1] [outbound] delivered file=1472505160080_1_5187_8878.relay domain="--redacted--.com" host="staff.--redacted--.com" ip=cli.ent.ip.addr port=25 mode=SMTP tls=N auth=N response="" delay=7230.286 fails=2 rcpts=1/0/0
[ERROR] [-] [core] [outbound] [outbound::25:cli.ent.ip.addr:undefined:50] release called twice for the same resource: Error
    at Pool.release (/usr/lib/node_modules/Haraka/node_modules/generic-pool/lib/generic-pool.js:409:65)
    at release_client (/usr/lib/node_modules/Haraka/outbound.js:1228:10)
    at finish_processing_mail (/usr/lib/node_modules/Haraka/outbound.js:1477:9)
    at pluggableStream.<anonymous> (/usr/lib/node_modules/Haraka/outbound.js:1709:25)
    at emitOne (events.js:77:13)
    at pluggableStream.emit (events.js:169:7)
    at pluggableStream.socket.process_data (/usr/lib/node_modules/Haraka/line_socket.js:23:20)
    at pluggableStream.<anonymous> (/usr/lib/node_modules/Haraka/line_socket.js:34:48)
    at emitOne (events.js:77:13)
    at pluggableStream.emit (events.js:169:7)
[WARN] [-] [core] [outbound] Socket [outbound::25:cli.ent.ip.addr:undefined:50] in pool got FIN
[INFO] [-] [core] [outbound] [outbound::25:cli.ent.ip.addr:undefined:50] force destroying all objects
[CRIT] [-] [core] Error: Callback was already called.
[CRIT] [-] [core]     at /usr/lib/node_modules/Haraka/node_modules/async/lib/async.js:43:36
[CRIT] [-] [core]     at HMailItem.hmail.next_cb (/usr/lib/node_modules/Haraka/outbound.js:98:9)
[CRIT] [-] [core]     at /usr/lib/node_modules/Haraka/outbound.js:2131:15
[CRIT] [-] [core]     at FSReqWrap.oncomplete (fs.js:82:15)
[NOTICE] [-] [core] Shutting down
root@relay:#
```

--- a/outbound.js
+++ b/outbound.js
@@ -1252,7 +1252,7 @@ function release_client (socket, port, host, local_addr) {
         return;
     }
     socket.__acquired = false;
-    
+
     var pool_timeout = cfg.pool_timeout;
     var name = 'outbound::' + port + ':' + host + ':' + local_addr + ':' + pool_timeout;
     if (!(server.notes && server.notes.pool)) {

--- a/outbound.js
+++ b/outbound.js
@@ -1237,11 +1237,22 @@ function get_client (port, host, local_addr, is_unix_socket, callback) {
     if (pool.waitingClientsCount() >= cfg.pool_concurrency_max) {
         return callback("Too many waiting clients for pool", null);
     }
-    pool.acquire(callback);
+    pool.acquire(function (err, socket) {
+        if (err) return callback(err);
+        socket.__acquired = true;
+        callback(null, socket);
+    });
 };
 
 function release_client (socket, port, host, local_addr) {
     logger.logdebug("[outbound] release_client: " + host + ":" + port + " to " + local_addr);
+
+    if (!socket.__acquired) {
+        logger.logerror("Release an un-acquired socket. Stack: " + (new Error()).stack);
+        return;
+    }
+    socket.__acquired = false;
+    
     var pool_timeout = cfg.pool_timeout;
     var name = 'outbound::' + port + ':' + host + ':' + local_addr + ':' + pool_timeout;
     if (!(server.notes && server.notes.pool)) {

--- a/outbound.js
+++ b/outbound.js
@@ -70,13 +70,16 @@ exports.load_config = function () {
         cfg.maxTempFailures = config.get('outbound.maxTempFailures') || 13;
     }
     if (!cfg.concurrency_max) {
-        cfg.concurrency_max = config.get('outbound.concurrency_max') || 100;
+        cfg.concurrency_max = config.get('outbound.concurrency_max') || 10000;
     }
     if (!cfg.connect_timeout) {
         cfg.connect_timeout = 30;
     }
     if (cfg.pool_timeout === undefined) {
         cfg.pool_timeout = 50;
+    }
+    if (!cfg.pool_concurrency_max) {
+        cfg.pool_concurrency_max = 10;
     }
     if (!cfg.ipv6_enabled && config.get('outbound.ipv6_enabled')) {
         cfg.ipv6_enabled = true;
@@ -1191,6 +1194,9 @@ function get_pool (port, host, local_addr, is_unix_socket, connect_timeout, pool
                     callback("Outbound connection timed out to " + host + ":" + port, null);
                 });
             },
+            validate: function(socket) {
+                return socket.writable;
+            },
             destroy: function(socket) {
                 logger.logdebug('[outbound] destroying pool entry for ' + host + ':' + port);
                 // Remove pool object from server notes once empty
@@ -1203,7 +1209,8 @@ function get_pool (port, host, local_addr, is_unix_socket, connect_timeout, pool
                     logger.logwarn("[outbound] Socket got an error while shutting down: " + err);
                 });
                 if (!socket.writable) return;
-                socket.send_command('QUIT');
+                logger.logprotocol("C: QUIT");
+                socket.write("QUIT\r\n");
                 socket.end(); // half close
                 socket.once('line', function (line) {
                     // Just assume this is a valid response
@@ -1211,7 +1218,7 @@ function get_pool (port, host, local_addr, is_unix_socket, connect_timeout, pool
                     socket.destroy();
                 });
             },
-            max: max || 1000,
+            max: max || 10,
             idleTimeoutMillis: pool_timeout * 1000,
             log: function (str, level) {
                 if (/this._availableObjects.length=/.test(str)) return;
@@ -1226,7 +1233,10 @@ function get_pool (port, host, local_addr, is_unix_socket, connect_timeout, pool
 
 // Get a socket for the given attributes.
 function get_client (port, host, local_addr, is_unix_socket, callback) {
-    var pool = get_pool(port, host, local_addr, is_unix_socket, cfg.connect_timeout, cfg.pool_timeout, cfg.concurrency_max);
+    var pool = get_pool(port, host, local_addr, is_unix_socket, cfg.connect_timeout, cfg.pool_timeout, cfg.pool_concurrency_max);
+    if (pool.waitingClientsCount() >= cfg.pool_concurrency_max) {
+        return callback("Too many waiting clients for pool", null);
+    }
     pool.acquire(callback);
 };
 

--- a/outbound.js
+++ b/outbound.js
@@ -1514,7 +1514,12 @@ HMailItem.prototype.try_deliver_host_on_socket = function (mx, host, port, socke
         return send_command('MAIL', 'FROM:' + self.todo.mail_from);
     };
 
+    var fp_called = false;
     var finish_processing_mail = function (success) {
+        if (fp_called) {
+            return self.logerror("finish_processing_mail called multiple times! Stack: " + (new Error()).stack);
+        }
+        fp_called = true;
         if (fail_recips.length) {
             self.refcount++;
             exports.split_to_new_recipients(self, fail_recips, "Some recipients temporarily failed", function (hmail) {

--- a/outbound.js
+++ b/outbound.js
@@ -1351,6 +1351,7 @@ HMailItem.prototype.try_deliver_host_on_socket = function (mx, host, port, socke
 
     socket.once('close', function () {
         if (processing_mail) {
+            self.logerror("Remote end " + host + ":" + port + " closed connection while we were processing mail. Trying next MX.");
             processing_mail = false;
             release_client(socket, port, host, mx.bind);
             return self.try_deliver_host(mx);

